### DISCUSSION TITLE
Should developers be able to 'opt out' of built-in capabilities, and if so, how?

### DISCUSSION THREAD

**Discussion by aaronsteers** [meltano/sdk#1071] [General]
We should document a way to opt out of capabilities that otherwise would be inferred by default.

Full list here: https://sdk.meltano.com/en/latest/capabilities.html

These all seem pretty low-priority to opt out of:

- `discover` - Theoretically developers could opt out of this if the catalog cannot be discovered ahead of time.
- `catalog` - The SDK handles this automatically. A developer could opt out, but behavior may be unpredictable.
- `state` - Perhaps if _no_ stream in the tap supports incremental or log-based replication, a developer may want to opt out, but there's also no harm in accepting the file, even if it does not contain bookmark values for any resumable streams.
- `test` - Again, this is implemented internally, so there should be little reason to opt out of it.
- `batch` - Batch message handling is automatic on both sides (even if there's no native implementation), so there's no strong reason for developers to opt out of it.

These are not automatically enabled, so there's no need to opt out:

- `log-based`.- Really not an issue as of now, since the SDK doesn't have support for this yet.
- `properties` - Deprecated, so there's no reason to prefer this.

This one should rightfully be impossible to opt out of:

- `about`

Presumably, some developers could prefer to opt out of this, although we haven't as of yet received any request to do so:

- `stream-maps`
- `record-flattening`

**Comment by edgarrmondragon** [comment]
Since part of the motivation for this discussion was the output of `--help` including options that aren't actually supported by the tap, specially state, an old PR (https://github.com/meltano/sdk/pull/936) might be relevant.

Other than that, I think it makes sense to have class-level attributes for each for these in the appropriate plugin classes, perhaps through a `Meta` nested class (similar to [Django](https://docs.djangoproject.com/en/4.1/topics/db/models/#meta-options)'s):

```python
class Tap:
  class Meta:
    supports_discovery = True
    supports_catalog = True
    supports_state = False  # none of the streams use incremental replication
    supports_batch = True
 	supports_log_replication = False
    supports_stream_maps = True
    supports_record_flattening = True
```
